Sources/NIO/SelectorGeneric.swift

```diff
@@ -0,0 +1,376 @@
+        guard self.lifecycleState == .open else {
+            throw IOError(errnoCode: EBADF, reason: "can't register on selector as it's \(self.lifecycleState).")
+        }
+
+        try selectable.withUnsafeHandle { fd in
+            assert(registrations[Int(fd)] == nil)
+            try self._register(selectable : selectable, fd: Int(fd), interested: interested)
+            registrations[Int(fd)] = makeRegistration(interested)
+        }
+    }
+
+    /// Re-register `Selectable`, must be registered via `register` before.
+    ///
+    /// - parameters:
+    ///     - selectable: The `Selectable` to re-register.
+    ///     - interested: The `SelectorEventSet` in which we are interested and want to be notified about.
+    func reregister<S: Selectable>(selectable: S, interested: SelectorEventSet) throws {
+        assert(self.myThread == NIOThread.current)
+        guard self.lifecycleState == .open else {
+            throw IOError(errnoCode: EBADF, reason: "can't re-register on selector as it's \(self.lifecycleState).")
+        }
+        assert(interested.contains(.reset), "must register for at least .reset but tried registering for \(interested)")
+        try selectable.withUnsafeHandle { fd in
+            var reg = registrations[Int(fd)]!
+            try self._reregister(selectable : selectable, fd: Int(fd), oldInterested: reg.interested, newInterested: interested)
+            reg.interested = interested
+            registrations[Int(fd)] = reg
+        }
+    }
+
+    /// Deregister `Selectable`, must be registered via `register` before.
+    ///
+    /// After the `Selectable is deregistered no `SelectorEventSet` will be produced anymore for the `Selectable`.
+    ///
+    /// - parameters:
+    ///     - selectable: The `Selectable` to deregister.
+    func deregister<S: Selectable>(selectable: S) throws {
+        assert(self.myThread == NIOThread.current)
+        guard self.lifecycleState == .open else {
+            throw IOError(errnoCode: EBADF, reason: "can't deregister from selector as it's \(self.lifecycleState).")
+        }
+        // temporary workaround to stop us delivering outdated events
+        self.deregistrationsHappened = true
+        try selectable.withUnsafeHandle { fd in
+            guard let reg = registrations.removeValue(forKey: Int(fd)) else {
+                return
+            }
+            try self._deregister(selectable: selectable, fd: Int(fd), oldInterested: reg.interested)
+        }
+    }
+
+    /// Apply the given `SelectorStrategy` and execute `body` once it's complete (which may produce `SelectorEvent`s to handle).
+    ///
+    /// - parameters:
+    ///     - strategy: The `SelectorStrategy` to apply
+    ///     - body: The function to execute for each `SelectorEvent` that was produced.
+    func whenReady(strategy: SelectorStrategy, _ body: (SelectorEvent<R>) throws -> Void) throws -> Void {
+        try _whenReady(strategy: strategy, body)
```

Comment:
```suggestion
        try self._whenReady(strategy: strategy, body)
```

NIO has mandatory `self.`